csharp/src/Ice/MetricsObserver.cs

```diff
@@ -0,0 +1,230 @@
+//
+// Copyright (c) ZeroC, Inc. All rights reserved.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using ZeroC.Ice;
+
+namespace ZeroC.IceMX
+{
+    public partial class Metrics
+    {
+        internal object Clone() => MemberwiseClone();
+    }
+    internal class MetricsHelper<T> where T : Metrics
+    {
+        private readonly AttributeResolver _attributes;
+
+        public virtual void InitMetrics(T metrics)
+        {
+            // Override in specialized helpers.
+        }
+
+        internal string Resolve(string attribute) => _attributes.Resolve(this, attribute);
+        protected MetricsHelper(AttributeResolver attributes) => _attributes = attributes;
+
+        protected virtual string? DefaultResolve(string attribute) => null;
+
+        internal class AttributeResolver
+        {
+            private readonly Dictionary<string, Func<object, object?>> _attributes =
+                new Dictionary<string, Func<object, object?>>();
+
+            public void Add(string name, Func<object, object?> resolver) => _attributes.Add(name, resolver);
+
+            public string Resolve(MetricsHelper<T> helper, string attribute)
+            {
+                if (!_attributes.TryGetValue(attribute, out Func<object, object?>? resolver))
+                {
+                    if (attribute.Equals("none"))
+                    {
+                        return "";
+                    }
+                    return helper.DefaultResolve(attribute) ?? throw new MissingFieldException(attribute);
```

Comment:
The impl of DefaultResolve currently throw ArgumentOutOfRange exception and don't return null... except for the base method implementation above. We should fix the implementations of DefaultResolve to either return null or throw and fix the code here accordingly.

Reply:
fixed